Unit test for messages controller

tokuchan-pg opened the pull request:
# WHAT
messagesコントローラーのテスト。indexアクションとcreateアクションそれぞれついて、以下の2つのテストを実施
- インスタンス変数が正しくセットされているかどうか
- ビューに正しく遷移するかどうか

# WHY
バグの無いサービスをリリースするため

## 設定についての補足
- gem 'rails-controller-testing'をインストールし、Rails5でもassignsを使ってテストできるように設定
- ControllerMacrosにより、ログイン認証された状態でテストを行えるように設定

## 行った４つのテスト

1. GET #indexアクション
インスタンス変数@messageが正しくセットされるかどうか
（生成されたインスタンスが、Messageクラスのインスタンスであるかどうかでテスト）

1. GET #indexアクション
GET #indexアクションに対応したビューが呼び出されているかどうか

1. POST #createアクション
ストロングパラメーターが入った状態のインスタンス変数@messageが正しくセットされるかどうか
（各カラムの中身をそれぞれ付け合わせることでテスト）

1. POST #createアクション
メッセージの保存に成功した時、messages#indexにリダイレクトされるかどうか

## メッセージコントローラーの状況
![2017-04-27 19 09 12](https://cloud.githubusercontent.com/assets/25572309/25478709/2f140472-2b7d-11e7-8dcc-4e51b7167815.png)




## テストの実行結果
![2017-04-27 18 52 40](https://cloud.githubusercontent.com/assets/25572309/25478075/c1cf06fc-2b7a-11e7-8064-cd8a1de9f921.png)

tokuchan-pg committed:
installed gem 'rails-controller-testing'

We need this gem to use 'assign' in controller testing from Rails5.

At the same time, set configuration in rails_helper.rb by referring to
below READE.me.

https://github.com/rails/rails-controller-testing/blob/master/README.md

tokuchan-pg committed:
generated ControllerMacros module for testing in log_in enviroment

tokuchan-pg committed:
tests for messages controller

Tested below 4 patterns

*GET #index
whether it assigns the new message to @message.

*GET #index
whether it renders the :index template

*POST #create
whether it assigns the new message to @message with strong parameter

*POST #index
whether it redirects to messages#index

KoichiOzaki commented:
スクリーンショットいつも丁寧に貼ってくださるのでレビューしやすいです＼(^o^)／
是非これからもレビューする人の立場にたって
・何をレビューしてほしいのか、レビューする際に必要な情報は何か
などを考えながらPRを作成してください！

KoichiOzaki commented on spec/controllers/messages_controller_spec.rb:
```diff
@@ -0,0 +1,45 @@
+      it "redirects to messages#index" do
+        post :create, params:{ group_id: group.id, message: attributes_for(:message)}
+        expect(response).to redirect_to group_messages_path
+      end
```
保存が失敗した時のテストもしましょう。
プロジェクトによってテストの粒度は違いますが、
一般的には最低限、全ての処理が1度は走るようにテストを記述したほうが良いです。

tokuchan-pg replied:
対応いたしました。

KoichiOzaki commented on spec/factories/messages.rb:
```diff
@@ -1,10 +1,21 @@
+    name      { Faker::Name.name }
+    email     { Faker::Internet.email }
+    password  { Faker::Internet.password }
+  end
```
モデルごとにファイルを分けましょう。

tokuchan-pg replied:
対応いたしました。

KoichiOzaki commented on spec/factories/messages.rb:
```diff
@@ -1,10 +1,21 @@
 
+  factory :group do
+    name      { Faker::Name.name }
+  end
```
ここもuserモデルと同じく、ファイルを分けましょう。

tokuchan-pg replied:
対応いたしました。

KoichiOzaki commented on spec/controllers/messages_controller_spec.rb:
```diff
@@ -0,0 +1,45 @@
+      it "renders the :index template" do
+        get :index, params:{ group_id: group.id }
+        expect(response).to render_template :index
+      end
```
set_group,set_messagesメソッドでインスタンス変数を作成してると思うので、その部分もテストしましょう。

tokuchan-pg replied:
テストを追加いたしました。

tokuchan-pg committed:
separeted factories files

tokuchan-pg committed:
added more controller tests

Added below 3 patterns

*GET #index & POST #create
whether it assigns the requested group to @group.

*GET #index & POST #create
whether it assigns the requested messages to @messages.

*POST #create
whether it renders to messages#index if creating a message failed

tokuchan-pg commented:
## 追加で実施した３つのテスト（合計追加数は５）

※1と2はmessagesコントローラーにてbefore_actionで呼び出しているprivate以下のメソッドのテスト

1. GET #indexアクション & POST #createアクション
インスタンス変数@groupが正しくセットされるかどうか

1. GET #indexアクション & POST #createアクション
インスタンス変数@messagesが正しくセットされるかどうか
（配列なので、復数のデータを生成し、matchマッチャでテストを行った）

1. POST #createアクション
メッセージの保存に失敗した時、messages#indexにrenderされるかどうか

## テストの実行結果
![2017-04-27 22 04 09](https://cloud.githubusercontent.com/assets/25572309/25484870/a55eff2e-2b96-11e7-80bf-209b076d4490.png)

KoichiOzaki commented:
[![LGTM](http://lgtm.in/p/I6y4IF9cq)](http://lgtm.in/i/I6y4IF9cq)
